### ISSUE TITLE
[ud] Release v1.3: multi-word tokens

### ISSUE THREAD

**Issue by makazhan** [ftyers/docs#13]
**Original e-mail:**
- What are the reasons for treating stuff like "барған жоқ" or "[ұл да] [қыз да]" as MWTs?
- Some MWT, e.g. "не болмаса" (cnjcoo), "сонда да" (cnjadv), have no detailed analyses (separate for не and болмаса). Is it by mistake or design?

We treat "барған жоқ/емес" as VERB and AUX (that must be wrong) linked by NEG relation.
We treat stuff like "не болмаса", "сонда да", etc. as pronouns (in case of не) and/or adpositions linked by MWE relation.

**Jonathan's reply:**

> - What are the reasons for treating stuff like "барған жоқ"

Our transducer treats this as a single negative past tense form.  I
think in Kazan we discussed changing this behaviour.  This is a major
change, though: if we do it, we'll have to change other
Turkic-language transducers, as well as pairs that rely on Kazakh.  So
it probably requires further discussion first.  For now we should just
worry about the best way to treat these in UD.  What do you propose?

> or "[ұл да] [қыз да]" as MWTs?

Really?  I blame Fran.  Both ‹да›s should be conjunctions, but I'm not
sure what type atm.

> - Some MWT, e.g. "не болмаса" (cnjcoo), "сонда да" (cnjadv), have no
>   detailed analyses (separate for не and болмаса). Is it by mistake or design?

By design.  Our analyser treats these as single conjunctions.  I think
we discussed these briefly in Kazan too?  The point is that while they
can be analysed more deeply, that different analysis implies a
different meaning.  Does this make sense to you, and if so, do you
agree with it?  I.e., as a native speaker of Kazakh, can you get two
different meanings out of such phrases depending on their analysis (or
use)?

> We treat "барған жоқ/емес" as VERB and AUX (that's must be wrong) linked by
> NEG relation.

Yeah, жоқ/емес are certainly not AUX, but there might be a way to
analyse them as separate words.  Let's discuss this one further.

> We treat stuff like "не болмаса", "сонда да", etc. as pronouns (in case of
> не) and/or adpositions linked by MWE relation.

Do you mean that you treat "да" as an adposition?  We treat it as
postadverb, I think.  Formally, they both follow nouns, but the
difference is that a noun should depend on an adposition, whereas "да"
depends on the noun.  Or at least this is my intuition; I could see
someone analysing it as a noun phrase depending on да, and the
resulting phrase as adverbial to the rest of the clause, in which case
it would be the same as an adposition.  Fran, I think we should
discuss this one further too.

**Comment by makazhan** [comment]
@jonorthwash 

> > What are the reasons for treating stuff like "барған жоқ"
> 
> Our transducer treats this as a single negative past tense form. I
> think in Kazan we discussed changing this behaviour. This is a major
> change, though: if we do it, we'll have to change other
> Turkic-language transducers, as well as pairs that rely on Kazakh. So
> it probably requires further discussion first. For now we should just
> worry about the best way to treat these in UD. What do you propose?

How about treating _барған_ as a verb in past tense and _жоқ_ as a particle that subordinates with relation neg. That's how _not_ is treated in the English UD. Same could be used for _емес_ in similar cases.
A side note: in such constructions _барған_ can also accept possession as in _Шын айтам! Барған**ым** жоқ. (I swear! I didn't go)_ I heard such usage only with px1sg though.

> > or "[ұл да] [қыз да]" as MWTs?
> 
> Really? I blame Fran. Both ‹да›s should be conjunctions, but I'm not
> sure what type atm.

Don't blame Fran :) That's my bad. Those aren't in the treebank.
Nevertheless, what's the reason to treat those as a single token?
Is it because they (poatadv and conj _да_) are indistinguishable during tokenization?

> > Some MWT, e.g. "не болмаса" (cnjcoo), "сонда да" (cnjadv), have no detailed analyses (separate for не and болмаса). Is it by mistake or design?
> 
> By design. Our analyser treats these as single conjunctions. I think
> we discussed these briefly in Kazan too? The point is that while they
> can be analysed more deeply, that different analysis implies a
> different meaning. Does this make sense to you, and if so, do you
> agree with it? I.e., as a native speaker of Kazakh, can you get two
> different meanings out of such phrases depending on their analysis (or
> use)?

Unfortunately, I am not really a native speaker. But as someone who learned the language as a 6-7 year old and spoke fairly regularly (for an urban person) since then, I perceive those as a 
single whole when they're used as conjuncts (subordinating or otherwise). Of course one can analyse _не болмаса_ as a subject verb pair or a copular predicate, but not when they comprise a conj. as a whole. Same for _сонда да_: neither conj. nor postadv seem to work for _да_, when the whole is used as a conjunct.
That's why I opted for the _multi-word expression_ relation. The point is that this way we can treat each word in isolation as whatever (e.g. не/PRON болмаса/CONJ) but the whole would function according to usage.
I propose similar treatment for stuff like _егер де_, _барақ та_ and few others. In those, _ДА_ is actually redundant or emphatic (cool word I learned today:)

> > We treat "барған жоқ/емес" as VERB and AUX (that's must be wrong) linked by
> > NEG relation.
> > Yeah, жоқ/емес are certainly not AUX, but there might be a way to
> > analyse them as separate words. Let's discuss this one further.

See my proposal above.

> > We treat stuff like "не болмаса", "сонда да", etc. as pronouns (in case of
> > не) and/or adpositions linked by MWE relation.
> 
> Do you mean that you treat "да" as an adposition? We treat it as
> postadverb, I think.

You see, in "classic" Kazakh grammars case-postpositions, conjuncts, postadv-s, particles, even question words (particles) were grouped under a common category _шылау_, and each received a sub-category, e.g. _септеулік шылау_ (case postpos.), _жалғаулық шылау_ (conjuncts), etc.
Stuff like _ДА_, _ҒАНА_, _-АҚ_ received the _демеулік шылау_ category, which I decided to label as adpositions for now and distinguish by surface form later if need be.
I regret that now:(

> Formally, they both follow nouns, but the
> difference is that a noun should depend on an adposition, whereas "да"
> depends on the noun. Or at least this is my intuition; I could see
> someone analysing it as a noun phrase depending on да, and the
> resulting phrase as adverbial to the rest of the clause, in which case
> it would be the same as an adposition.

I sort of understand the difference, but one of UD requirements is to always subordinate function words to the content ones (even case-adpositions). Without following this requirement, I'm afraid we won't be able to release the data, but in that case distinction with adpositions is not clear and we need separate pos label and/or dependency relation.

**Comment by jonorthwash** [comment]
> A side note: in such constructions барған can also accept possession as in _Шын айтам! Барғаным жоқ_. (_I swear! I didn't go_) I heard such usage only with px1sg though.

This is the normal negation in Tatar and I think Uzbek also, and they do it with all persons.

**Comment by ftyers** [comment]
About 

```
or "[ұл да] [қыз да]" as MWTs?
```

This is the same reason as for ғана in the issue #12 ... the underlying form is -DA so in order to properly generate we need to have them as a single unit.

**Comment by jonorthwash** [comment]
> How about treating _барған_ as a verb in past tense and _жоқ_ as a particle that subordinates with relation neg. That's how not is treated in the English UD.

The problem with doing this is that the agreement comes on the negative word, e.g. _барған жоқпын_.  I guess this is like English auxiliaries (_haven't gone_), but жоқ is very clearly not an auxiliary in Kazakh.  If you analyse _барған_ as a verb in past tense, then it would have to be person- and number-less, right?  This remains tricky :\

In _барғаным жоқ_ (a different construction), you can analyse this as _бар_`<v><iv><ger_past><px1sg><nom>` _жоқ_`<adj>`+_е_`<cop><aor><p3><sg>`.  I.e., treating them as separate tokens (plus one more) is the easiest approach.

**Comment by ftyers** [comment]
> I sort of understand the difference, but one of UD requirements is to always subordinate function words to the content ones (even case-adpositions). Without following this requirement, I'm afraid we won't be able to release the data, but in that case distinction with adpositions is not clear and we need separate pos label and/or dependency relation.

I agree with this, in both cases the content word should be head, but we can distinguish them in the dependency relation. Adpositions get `case`, while the other stuff can get `advmod`. At some point we could even give a sub-category for the postadv/mods (based on the lemma). In any case this is not going to be difficult for the parser to resolve.

**Comment by jonorthwash** [comment]
So with postpositions it would be like this?
(shorthand)

```
...
3. үйге үй n.dat noun ...
4. қарай қарай postpos case 3
```

And with postadv like this:

```
...
3. үйге үй n.dat noun ...
4. ғана ғана part advmod 3
```

It's kind of weird having a particle adverb-modding a noun...  But otherwise it's fairly consistent with what we do in our analyser, and unique enough that it'll be easy to identify and change if it's decided to do something different later.

**Comment by ftyers** [comment]
@jonorthwash Yep that looks good to me, and UD-compatible.
